### DISCUSSION TITLE
Login with custom token

### DISCUSSION THREAD

**Discussion by mars3142** [GodotNuts/GodotFirebase#189] [Ideas]
Firebase offers the login with custom token method. I want to use it, because I don't want to use email/password for login. The Firebase team offers a simple solution for that with custom tokens. But I can't use it with your plugin. Correct? 

It would be really nice to have such an option. Should I create a Feature Request for that?

PS: For more information about username/password login with Firebase see https://github.com/firebase/functions-samples/tree/master/username-password-auth

**Comment by WolfgangSenff** [comment]
We support a whole host of login providers, including OAuth, Google, and more. Feel free to go over the code here to understand it: https://github.com/GodotNuts/GodotFirebase/blob/main/addons/godot-firebase/auth/auth.gd

It is also documented in our wiki here: https://github.com/GodotNuts/GodotFirebase/wiki/Authentication-and-User-Management

Edit:

Okay, I'm correcting myself here. It seems we don't support it directly, though one of our team has offered to implement it soon. I'll create an issue. Thanks for pointing it out!

**Comment by fenix-hub** [comment]
Hi @mars3142 , I'm working on this feature and I've implemented the login via a custom token , even though it is not possible to make it work out of the box as the example you posted. Here's the reason why:
The implementation you shared is not actually using a real Firebase function to authenticate a user via username/password, but ([step 1](https://github.com/firebase/functions-samples/blob/58b09a4f6d914d877da0ecb04633a75ae72b1e30/username-password-auth/functions/index.js#L102)) is using another external API (https://httpbin.org/basic-auth) to check if a user/password combination is correct via an HTTPRequest, ([step 2](https://github.com/firebase/functions-samples/blob/58b09a4f6d914d877da0ecb04633a75ae72b1e30/username-password-auth/functions/index.js#L89)) and if it is the function will ask to the FirebaseSDK to create a custom token based on the username. ([step 3](https://github.com/firebase/functions-samples/blob/58b09a4f6d914d877da0ecb04633a75ae72b1e30/username-password-auth/functions/index.js#L90)) Once returned by firebase, the token is used to authenticate with the token.
While I'm able to easely implement the 3rd step using a Firebase Rest Endpoint, the issue consists in the 2nd step: creating a custom token.
Firebase requires an RS256 Token which can be created **only with** their SDK or using third party libraries (ref: https://firebase.google.com/docs/auth/admin/create-custom-tokens).
Thus, since we are not implementing their SDKs, and we are not currently planning to implement 3rd party libraries, the creation of that currently impossible to implement for us.
So we can implement the actual function to authenticate with a custom token, but we are not able to let you create that token, and it cannot be a random string.

However, I'll propose a few workarounds for this, and the second one of them is already working without any update required:
1) You can implement the 1st step by yourself to get a username/password combination check with an HTTP POST Request, pointing to `https://httpbin.org/basic-auth/{username}/{password}` and with a simple body such:
```
{
"auth" : {
 "username": [username],
 "password" : [password]
}
}
```
and based on the response (it contains a `true/false` `authenticated` field) you can then enable your own logic to authenticate to Firebase
2) Use the `Firebase.Auth.signup_with_email_and_password(email, password)` function using a fixed `email` variable:
```
var username : String = ""
var password : String = ""
const _EMAIL_SUFFIX = "@email.xyz"

# ... implement your own logic to get a username and password input from user

func _on_login():
  Firebase.Auth.singup_with_email_and_password(username+_EMAIL_SUFFIX, password)
```
Of course this will work if you enable email/password combination only, but this will basically work out of the box since there is no check from Firebase about the authenticity of an email.

Let me know what you think about it, if you have any suggestion or more information to share about a possible solution.
Eventually I could anyway implement the custom token Authorization method (but then you will have to create a token on your own), but *also* implement a "fake" singup/singin username/password method which uses internally a given, fixed or randomized email suffix, so you don't have to actually write an _EMAIL_SUFFIX variable and  you can call a function with a different name.

**Reply by mars3142** [reply]
https://github.com/firebase/functions-samples/blob/master/username-password-auth/functions/index.js#L89

This line create the RS256 token, but I don't know how to use this as a login into Firebase with your plugin. If I'm able to login with that token, my user will be authenticated within Firebase and be show in the authentication section in the Firebase Console.

Please show a login with that token instead of email/password. If I can use that token, I can implement username/email login (with the help of an external microservice for checking the login and creating the token). A login with added domain is bad practice. I also know, that I have to do the username/password check on my own.

**Reply by fenix-hub** [reply]
Thanks for your answer.
As you can see here #192 I've implemented a new method.
Just call the `login_with_custom_auth(token : String)` function and a request will be issued at the `https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken` endpoint (which should be similar to the one you also shared in the answer on the issue #191 )
I'm not able to test it, but anyway it should work as any other `login_*` function.
If you have any issue feel free to answer both here on on the issue.

**Reply by mars3142** [reply]
@fenix-hub I've tested your implementation. The login with the token is working, but the signal and automatic refresh of the token seems to be not implemented the right way. Should I try to fix it with a PR or will you do it (with the help of my test project)? If you want to fix it, please mail me, so I can send you the credentials.

**Reply by fenix-hub** [reply]
Hello @mars3142 ,
What do you mean by "the signal and automatic refresh of the token seems to not be implemented the right way"?
Is the signal not emitted/wrong? Is the refreshing failing/returning wrong values?
I'm at work right now and I'll be able to check the problem out only later today, or maybe tomorrow morning.
If you want you can try fixing it your way and making a PR that I can check out. I could give you some tips anyway to assist you indirectly.
I'll ask you for your credentials when I'll be able to work on my PC.
Let me know!! :)

**Reply by mars3142** [reply]
Hi @fenix-hub,

see #191, because I wrote my findings there. But I will try to create a PR for you to revisit.

**Reply by fenix-hub** [reply]
Thank you , and sorry for pinging back and forth between discussions and issues. Would be useful if github would give the chance to link them in some ways :P
